### PR TITLE
mgr/dashboard: Update Angular version to 12 

### PR DESCRIPTION
**Pull request by nizamial09**
There was a bug from angular 12 which affected us.

1. [Removing emitDecoratorMetadata Breaks Jest Config](https://github.com/angular/angular-cli/issues/20823)
Fixed by keeping the `emitDecoratorMetadata` in the `tsconfig.json` file

A full changelog can be seen here: https://blog.angular.io/angular-v12-is-now-available-32ed51fbfd49

For us, the most I had to do is to take care of the min-max validation
and a small CSS change regarding the math()

Depends on https://github.com/ceph/ceph-build/pull/1932
Fixes: https://tracker.ceph.com/issues/53049
Signed-off-by: Nizamudeen A <nia@redhat.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by nizamial09** [comment]
jenkins test dashboard

**Comment by nizamial09** [comment]
jenkins test dashboard

**Comment by nizamial09** [comment]
jenkins test dashboard cephadm

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/telemetry/telemetry.component.ts [hunk 1]
```diff
@@ -118,16 +117,6 @@ export class TelemetryComponent extends CdForm implements OnInit {
...
-      case 'int':
-        result.push(CdValidators.number());
-        result.push(Validators.required);
-        if (_.isNumber(option.min)) {
-          result.push(Validators.min(option.min));
-        }
-        if (_.isNumber(option.max)) {
-          result.push(Validators.max(option.max));
-        }
-        break;
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/telemetry/telemetry.component.ts:130]: this is an incorrect removal, i'll have to take care of it. so I'll put DNM label and will take care of this once I am back

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Comment by avanthakkar** [comment]
jenkins test api

**Comment by alfonsomthd** [comment]
jenkins test dashboard cephadm

**Review by avanthakkar — APPROVED** [review]
Tested locally and nothing breaking.

File: src/pybind/mgr/dashboard/frontend/angular.json [hunk 2]
```diff
@@ -75,7 +80,11 @@
                   "maximumWarning": "6kb"
                 }
               ],
-              "optimization": true,
+              "optimization": {
+                "styles": {
+                    "inlineCritical": false
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/angular.json:85]: What's the reason for this change? is there any related issue that can be reproducible?

File: src/pybind/mgr/dashboard/frontend/package.json [hunk 3]
```diff
@@ -99,28 +99,28 @@
...
+    "ngx-pipe-function": "1.0.0",
     "ngx-toastr": "13.1.0",
     "rxjs": "6.6.3",
     "simplebar-angular": "2.3.0",
-    "swagger-ui": "3.49.0",
-    "tslib": "^2.0.0",
-    "zone.js": "0.10.3"
+    "swagger-ui": "3.52.5",
+    "tslib": "2.3.1",
+    "zone.js": "~0.11.4"
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/package.json:108]:
```suggestion
    "zone.js": "0.11.4"
```

**Review by alfonsomthd — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard/frontend/angular.json (hunk 2 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/angular.json:85]: there was. the one I mentioned in the description. https://github.com/angular/angular-cli/issues/20760 Although now its seems to be resolved (atleast closed). I'll test it again.

File: src/pybind/mgr/dashboard/frontend/package.json (hunk 3 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/package.json:108]: This seems to be from running the `ng update`. I'll make the change, thanks

File: src/pybind/mgr/dashboard/frontend/angular.json [hunk 4]
```diff
@@ -34,7 +34,6 @@
               "file-saver",
               "lodash"
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/angular.json:35]: @nizamial09 I know that this is not strictly related to this PR, but could you kindly add at the beginning of this list
`brace-expansion` in order to avoid this warning?
```
Warning: /ceph/src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts depends on 'brace-expansion'. CommonJS or AMD dependencies can cause optimization bailouts.
```

File: src/pybind/mgr/dashboard/frontend/angular.json (hunk 2 above)
Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/frontend/angular.json:85]: @nizamial09 Thanks. i tested the bundle generation and it was successful:
```
✔ Browser application bundle generation complete.
✔ Localized bundle generation complete.
✔ Copying assets complete.
✔ Index html generation complete.
```

File: src/pybind/mgr/dashboard/frontend/angular.json (hunk 4 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/angular.json:35]: Yup. I'll do it. @alfonsomthd

File: src/pybind/mgr/dashboard/frontend/angular.json (hunk 2 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/angular.json:85]: Perfect! Then I'll take care of it too. Thanks @alfonsomthd

Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/angular.json]: @alfonsomthd today I made the change and repushed. I noticed that this is failing on dashboard-cephadm job. So my guess is in dev environment it this error is not happening since optimization is false but on prod environment this is happening.
```
✔ Browser application bundle generation complete.
✔ Browser application bundle generation complete.
- Generating localized bundles...
✔ Localized bundle generation complete.
- Copying assets...
✔ Copying assets complete.
- Generating index html...
✖ Index html generation failed.
undefined:2:146736: property missing ':'
```

Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/angular.json]: https://jenkins.ceph.com/job/ceph-dashboard-cephadm-e2e/1338/console

Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/frontend/angular.json:86]: @nizamial09 After investigating the issue I found that [upgrading to ngx-toastr v14.1.2 fixes the issue](https://github.com/scttcper/ngx-toastr/pull/901):
```
Option "--prod" is deprecated: Use "--configuration production" instead.
✔ Browser application bundle generation complete.
✔ Localized bundle generation complete.
✔ Copying assets complete.
⠋ Generating index html...1 rules skipped due to selector errors:
  .custom-file-input:lang(en)~.custom-file-label -> unmatched pseudo-class :lang
✔ Index html generation complete.
```
Can you confirm that the upgrade does not have any impact?

File: src/pybind/mgr/dashboard/frontend/package.json [hunk 5]
```diff
@@ -99,28 +99,28 @@
     "ng-block-ui": "3.0.2",
     "ng-click-outside": "7.0.0",
     "ng2-charts": "2.4.2",
-    "ngx-pipe-function": "^1.0.0",
+    "ngx-pipe-function": "1.0.0",
     "ngx-toastr": "13.1.0",
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/package.json:103]:
```suggestion
    "ngx-toastr": "14.1.4",
```
Screenshot from 2021-11-12 10-26-52

**Review by alfonsomthd — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard/frontend/angular.json (hunk 2 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/angular.json:86]: Oh okay. I'll test it out. Thanks @alfonsomthd

Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/angular.json:86]: just noticed the error `.custom-file-input:lang(en)~.custom-file-label -> unmatched pseudo-class :lang`

File: src/pybind/mgr/dashboard/frontend/package.json [hunk 6]
```diff
@@ -75,15 +75,15 @@
...
+    "@angular/common": "12.2.8",
+    "@angular/compiler": "12.2.8",
+    "@angular/core": "12.2.8",
+    "@angular/forms": "12.2.8",
+    "@angular/localize": "12.2.8",
+    "@angular/platform-browser": "12.2.8",
+    "@angular/platform-browser-dynamic": "12.2.8",
+    "@angular/router": "12.2.8",
     "@circlon/angular-tree-component": "10.0.0",
     "@ng-bootstrap/ng-bootstrap": "7.0.0",
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/package.json:88]:
```suggestion
    "@ng-bootstrap/ng-bootstrap": "10.0.0",
```
It seems we need to [upgrade the version](https://github.com/ng-bootstrap/ng-bootstrap/releases/tag/10.0.0).

**Review by alfonsomthd — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard/frontend/package.json (hunk 6 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/package.json:88]: Ah yes, i'll take care of that here. Also would it make sense to go through most of the npm packages and upgrade them (not everything but atleast some of the important one) in a different PR? @alfonsomthd

Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/frontend/package.json:88]: In this PR we should upgrade only those that are required in order to work properly with Angular 12.

File: src/pybind/mgr/dashboard/frontend/angular.json (hunk 2 above)
Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/frontend/angular.json:86]: This warning is related to the bootstrap version 4 (it foes not happen with v5),
but upgrading bootstrap is out of scope of this PR.

**Comment by alfonsomthd** [comment]
@nizamial09 It seems the last v12 patch version is v12.2.13:
```
-    "@angular/localize": "12.2.8",
-    "@angular/platform-browser": "12.2.8",
-    "@angular/platform-browser-dynamic": "12.2.8",
-    "@angular/router": "12.2.8",
+    "@angular/animations": "12.2.13",
+    "@angular/common": "12.2.13",
+    "@angular/compiler": "12.2.13",
+    "@angular/core": "12.2.13",
+    "@angular/forms": "12.2.13",
+    "@angular/localize": "12.2.13",
+    "@angular/platform-browser": "12.2.13",
+    "@angular/platform-browser-dynamic": "12.2.13",
+    "@angular/router": "12.2.13",
[...]
   "devDependencies": {
-    "@angular-devkit/build-angular": "12.2.7",
-    "@angular/cli": "12.2.7",
-    "@angular/compiler-cli": "12.2.8",
-    "@angular/language-service": "12.2.8",
+    "@angular-devkit/build-angular": "12.2.13",
+    "@angular/cli": "12.2.13",
+    "@angular/compiler-cli": "12.2.13",
+    "@angular/language-service": "12.2.13",
```

File: src/pybind/mgr/dashboard/frontend/angular.json (hunk 2 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/angular.json:86]: @alfonsomthd I've upgraded the ngx-toaster and also updated the ng-bootstrap to 10 and noticed this error is not there. I've also updated the minor version of angular using `npx-update` command
```
✔ Browser application bundle generation complete.
✔ Browser application bundle generation complete.
- Generating localized bundles...
✔ Localized bundle generation complete.
- Copying assets...
✔ Copying assets complete.
- Generating index html...
✔ Index html generation complete.
```

File: src/pybind/mgr/dashboard/frontend/package.json (hunk 5 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/package.json]: done!

File: src/pybind/mgr/dashboard/frontend/package.json (hunk 6 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/package.json]: Done and created a [tracker for npm upgrades ](https://tracker.ceph.com/issues/53274)

**Comment by nizamial09** [comment]
> @nizamial09 It seems the last v12 patch version is v12.2.13:

@alfonsomthd I updated to the latest patch version now.

File: src/pybind/mgr/dashboard/frontend/package.json (hunk 6 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/package.json]: ng-bootstrap is causing some issues with unit-testing, so I'll fix that and update

**Comment by nizamial09** [comment]
Upgrading ng-bootstrap version to 10 causes this bug: https://github.com/ceph/ceph/pull/40618#pullrequestreview-629010639

Also a translation failure happens in the make check
```
The following source translations in 'messages.xlf' need to be   fixed, please check the I18N suggestions on https://docs.ceph.com/en/latest/dev/developer_guide/dash-devel/#i18n:

13           <source> Slide <x id="INTERPOLATION" equiv-text="OnHover$ = new"/> of <x id="INTERPOLATION_1" equiv-text=";
14     private _pa"/> </source>
15           <context-group purpose="location">
16             <context context-type="sourcefile">node_modules/src/carousel/carousel.ts</context>
17             <context context-type="linenumber">114,118</context>
18           </context-group>
```

Animations are enabled globally and it causes some minor unit test failures. I'll have to find a way to globally disable the animations only in the testing framework. Currently I handled the failure like [this](https://github.com/ceph/ceph/pull/43404/files#diff-98e8ed354b14bf2cbceacd4e1d35d72485ab08f63d7ca903f30621815d4301f0R665)

File: src/pybind/mgr/dashboard/frontend/package.json (hunk 6 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/package.json]: Done!

**Comment by nizamial09** [comment]
jenkins test dashboard cephadm

**Comment by nizamial09** [comment]
jenkins test api

**Comment by nizamial09** [comment]
api test failure: https://jenkins.ceph.com/job/ceph-api/28973/

**Comment by nizamial09** [comment]
jenkins test api

File: doc/dev/developer_guide/dash-devel.rst [hunk 7]
```diff
@@ -215,8 +215,8 @@ The build process is based on `Node.js <https://nodejs.org/>`_ and requires the
 Prerequisites
 ~~~~~~~~~~~~~
 
- * Node 10.0.0 or higher
- * NPM 5.7.0 or higher
+ * Node 12.14.0 or higher
```
Comment (alfonsomthd) [inline doc/dev/developer_guide/dash-devel.rst:218]: Shouldn't this be aligned with the real version we're using?
https://github.com/ceph/ceph/blob/master/make-dist#L135

Comment (nizamial09) [reply doc/dev/developer_guide/dash-devel.rst:218]: You are right, I'll make the change

File: src/pybind/mgr/dashboard/run-frontend-unittests.sh [hunk 8]
```diff
@@ -31,7 +31,7 @@ if [ $? -gt 0 ]; then
   echo -e "\nTranslations extraction has failed."
 else
   i18n_lint=`awk '/<source> |<source>$| <\/source>/,/<\/context-group>/ {printf "%-4s ", NR; print}' src/locale/messages.xlf`
-  if [ "$i18n_lint" ]; then
+  if [ "$i18n_lint" ] && [[ $i18n_lint != *"node_modules/"* ]]; then
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/run-frontend-unittests.sh:34]: IMHO the conditional checks can be simplified .

**Review by alfonsomthd — APPROVED** [review]
Tested locally: LGTM

File: src/pybind/mgr/dashboard/run-frontend-unittests.sh (hunk 8 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/run-frontend-unittests.sh]: Done!

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.scss [hunk 9]
```diff
@@ -10,6 +10,7 @@
 
 .dataTables_wrapper {
   margin-bottom: 25px;
+  max-width: 99.9%;
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.scss:13]: Why 99.9% and not 100%? Maybe worth adding some comment explaining such a tricky value.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-form/rgw-user-form.component.ts [hunk 10]
```diff
@@ -107,11 +107,7 @@ export class RgwUserFormComponent extends CdForm implements OnInit {
       max_buckets_mode: [1],
       max_buckets: [
         1000,
-        [
-          CdValidators.requiredIf({ max_buckets_mode: '1' }),
-          CdValidators.number(false),
-          Validators.min(1)
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-form/rgw-user-form.component.ts:113]: Why was 1 here and 0 in the template?

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/alert-panel/alert-panel.component.html [hunk 11]
```diff
@@ -1,6 +1,6 @@
 <ngb-alert type="{{ bootstrapClass }}"
            [dismissible]="dismissible"
-           (close)="onClose()">
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/components/alert-panel/alert-panel.component.html:3]: Is this a change in the ngb-alert API?

**Review by epuertat — APPROVED** [review]
LGTM @nizamial09! Thank you. With these PRs that involve so many unrelated changes to different componets, it'd be great to quickly summarize the reasons and decisions made.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-form/rgw-user-form.component.ts (hunk 10 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-form/rgw-user-form.component.ts:113]: Not exactly. Its 1 in the template too 
https://github.com/ceph/ceph/blob/d53261a5b050b880002094099dbd5e85b3218631/src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-form/rgw-user-form.component.html#L134-L138

The change I am making here is for `user_quota_max_objects` and `bucket_quota_max_objects`. But thanks for pointing it out, I'll need to remove the min validator here also.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.scss (hunk 9 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.scss:13]: It was a hack to prevent [this issue ](https://github.com/ceph/ceph/pull/40618#pullrequestreview-629010639). I couldn't find an alternate for this.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/alert-panel/alert-panel.component.html (hunk 11 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/components/alert-panel/alert-panel.component.html:3]: After bootstrap 8.0.0 I saw this behavior is changed and close was no longer working. https://ng-bootstrap.github.io/#/components/alert/api

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.scss (hunk 9 above)
Comment (epuertat) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.scss:13]: Yes, please. So the issue was already there... 🤔

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/alert-panel/alert-panel.component.html (hunk 11 above)
Comment (epuertat) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/components/alert-panel/alert-panel.component.html:3]: Ah, they changed the output `close` with `closed`, and now `close` is a method to close the alert... Thanks!

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-form/rgw-user-form.component.ts (hunk 10 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-form/rgw-user-form.component.ts:113]: Done!

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.scss (hunk 9 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.scss:19]: After bootstrap 8, this seems to be happening. I am pretty sure its related to our implementation. I've tried debugging it a lot but couldn't find the cause. Added a comment.

**Comment by nizamial09** [comment]
jenkins test dashboard cephadm

**Comment by nizamial09** [comment]
jenkins test dashboard cephadm

**Review by votdev — APPROVED** [review]
LGTM
